### PR TITLE
ARROW-1756: [Python] Fix large file read/write error

### PR DESCRIPTION
**Pull request by Licht-T**
This is the part of [ARROW-1756](https://issues.apache.org/jira/browse/ARROW-1756).

### CODE REVIEW

File: cpp/src/arrow/io/file.cc
Comment (wesm) [inline cpp/src/arrow/io/file.cc]: Since `N` is never changed, just use `nbytes`? Maybe make it `const int64_t nbytes` in the function arguments for good measure

File: cpp/src/arrow/io/file.cc
Comment (wesm) [inline cpp/src/arrow/io/file.cc]: Use `*bytes_read` instead of `n`?

File: cpp/src/arrow/io/file.cc
Comment (wesm) [inline cpp/src/arrow/io/file.cc]: Use a more descriptive variable name than `i`

File: cpp/src/arrow/io/file.cc
Comment (wesm) [inline cpp/src/arrow/io/file.cc]: Similar comments as above.

File: python/pyarrow/tests/test_feather.py
Comment (wesm) [inline python/pyarrow/tests/test_feather.py]: Any data will do here, so you could do `np.arange` here to make this faster rather than generating so much random data

File: python/pyarrow/tests/test_feather.py
Comment (wesm) [inline python/pyarrow/tests/test_feather.py]: `['A']`

**Review by wesm — COMMENTED** [review]
Is this issue limited to Mac? I thought the bug reporter was on Linux, but either way this is a good thing to fix

**Comment by wesm** [comment]
Thanks for looking at this

**Comment by Licht-T** [comment]
@wesm I checked on Linux, but this only fixes the problem on Mac.
I am also trying to find the cause of the Linux issue.

**Comment by Licht-T** [comment]
@wesm I found the reason why.
I'll fix.
http://man7.org/linux/man-pages/man2/write.2.html

**Comment by wesm** [comment]
OK, thanks. I'll wait on you to address the code comments and then we can merge this

**Comment by Licht-T** [comment]
@wesm Fixed. Now it works fine on Linux!

**Review by xhochy — APPROVED** [review]
+1, thanks for taking a deep dive into those internals.

Will merge once the build is green.

**Comment by Licht-T** [comment]
@xhochy I want to solve the Windows issue as same.
Should I make a JIRA issue?

**Comment by xhochy** [comment]
@Licht-T yes, please open a JIRA.

I would merge this PR once Travis is happy but sadly C++ builds keep timeouting on Travis. I will re-trigger later again and merge on success.

File: cpp/src/arrow/io/file.cc
Comment (wesm) [inline cpp/src/arrow/io/file.cc]: Typos in these defines, I will fix

File: cpp/src/arrow/io/file.cc
Comment (wesm) [inline cpp/src/arrow/io/file.cc]: We should use descriptive variable names. I will fix

**Comment by wesm** [comment]
This is infinite-looping on Linux, debugging

**Comment by wesm** [comment]
+1

**Comment by wesm** [comment]
thanks @Licht-T!

**Pull request closed without merging**
